p2p/src/protocol/peermgr.rs

```diff
@@ -260,21 +270,55 @@ impl<U: Handshake + SetTimeout + Connect + Disconnect + Events> PeerManager<U> {
 
     /// Initialize the peer manager. Must be called once.
     pub fn initialize<A: AddressSource>(&mut self, time: LocalTime, addrs: &mut A) {
-        let retry = self
+        let peers = self
             .config
-            .retry
+            .persistent
             .iter()
             .take(self.config.target_outbound_peers)
             .cloned()
             .collect::<Vec<_>>();
 
-        for addr in retry {
-            self.connect(&addr, time);
+        for addr in peers {
+            if !self.connect(&addr, time) {
+                panic!(
+                    "{}: unable to connect to persistent peer: {}",
+                    source!(),
+                    addr
+                );
+            }
         }
         self.upstream.set_timeout(IDLE_TIMEOUT);
         self.maintain_connections(addrs, time);
     }
 
+    fn retrier_add_peer(&mut self, addr: &net::SocketAddr, local_time: LocalTime) {
+        let attempts = self.retry_attempts.entry(*addr).or_default();
+        let delay = LocalDuration::from_secs(2_u64.pow((*attempts).min(63)))
+            .clamp(self.config.retry_min_wait, self.config.retry_max_wait);
+        self.retry_at.insert(*addr, local_time + delay);
+        self.upstream.set_timeout(delay);
+        *attempts += 1;
+    }
+
+    fn retrier_remove_peer(&mut self, addr: &net::SocketAddr) {
+        debug_assert!(self.is_connected(addr));
+        self.retry_attempts.remove(addr);
+        self.retry_at.remove(addr);
+    }
+
+    fn retrier_reconnect(&mut self, local_time: LocalTime) {
+        let peers: Vec<_> = self
+            .retry_at
+            .iter()
+            .filter(|(_, v)| *v <= &local_time)
```

Comment:
This condition doesn't look right, shouldn't it be `v >= local_time`? ie. "get me all peers where the retry time is passed the current time?